BIP drafts: SwiftSync Specification (#2152)
bitcoin/bips: bip-0455.md

Inputs of a Bitcoin block are referenced by the outpoint data structure. This poses a limitation during initial block download (IBD), such that a client must process blocks sequentially to validate the chain history. The SwiftSync protocol allows blocks to be evaluated in arbitrary order, however additional data is required that must be served over the peer-to-peer network. Namely, the creation height, coinbase flag, input script, and amount for each spent coin must be accessible to fully validate a block in a state-less manner. This document describes how to share this additional data over the peer to peer network.
## Motivation

A current limitation of IBD is that it must be done sequentially. This is a result of the height, coinbase flag, input script, and amount of the block inputs being omitted from the data committed to by proof of work in the current block, and, thus, this data cannot be trusted if received over the wire naively. Using the SwiftSync protocol, a client is able to verify the correctness of this data, even if served by a potentially untrusted party. This allows a significant improvement in IBD performance, as block downloads may be done in parallel.

> ## Specification

Reviewer: Minor formatting thing, but the document as reads today isn't uniform w.r.t when it decides to add a new line before heading vs not.

#### Coin

| Field                    | Type                          | Serialization | Description                                                                                                       |
| :----------------------- | :---------------------------- | ------------- | ----------------------------------------------------------------------------------------------------------------- |
| Input index              | 32 bit unsigned integer       | Little endian | The index in the block inputs for which this coin corresponds. The coinbase inputs are _excluded_ from this index |
| Height and coinbase flag | Height code                   | Defined above |                                                                                                                   |
| Script                   | Reconstructable script format | Defined above |                                                                                                                   |
| Amount                   | 64 bit unsigned integer       | Defined above | Satoshi denominated value                                                                                         |
### Messages

> #### MSG_GET_SPENT_COINS

Reviewer: Is the idea that a peer would issue this request for _every_ block in the chain? If we assume mainnet at height, and a 150 ms round trip time, then a peer would spend nearly 80 hours just downloading this undo data. 

You may want to consider a batched variant, similar to the way messages like `getheaders` works.

Author: We've found that bandwidth throughput is the limiting factor when downloading blocks in parallel. Not all spent coins have to be downloaded if a client keeps a cache, as this document describes. In the batched variant, the cache is not possible and the bandwidth requirement increases significantly.

`MSG_GET_SPENT_COINS` defines a request for the inputs of a block.

Define `cmdString` as `getbspent`. Define BIP-324 message type as ???.

| Field       | Type                    | Description                                                               |
| :---------- | :---------------------- | ------------------------------------------------------------------------- |
| `version`   | `byte`                  | Version of the expected serialization in the response                     |
| `blockhash` | 32 byte vector          | Hash of the block for which inputs are requested                          |
| `cutoff`    | 32 bit unsigned integer | If set greater than zero, only include coins created _before_ this height |

#### MSG_SPENT_COINS

> `MSG_SPENT_COINS` defines the data structure for inputs of a block.

Reviewer: Probably not really y'all's intended use case, but if you optionally make it possible to include merkle proofs for the set of coins, then this message can be used to obtain a proof that an output was spent in a given block.

Reviewer: It would actually also be useful for BIP 157+158 peers, as the final version that shipped includes the script spent (instead of the outpoint), which means that if you're using the filters to find a block where a given script has been spent, you need to make some assumptions about what the prev script is for a given transaction.

Author: The most recent response on [this mailing list post](https://groups.google.com/g/bitcoindev/c/rThmyI8ZN3Q/m/TJBc1xRbAQAJ) mentions commitment to the UTXO set as part of the block header. There are additional ways to do this outside of a soft fork as well, i.e. utreexo proofs. For now I think it best to leave this unspecified in this version of the message while the community shares ideas, but I do think this is interesting.

```
  BIP: ?
  Layer: Peer Services
  Title: Peer sharing of block spent coins
  Authors: Robert Netzke <bips@2140.dev>
  Deputies: Ruben Somsen <bips@2140.dev>
  Status: Draft
  Type: Specification
  Assigned: ?

>   License: BSD-3-Clause

Reviewer 2: Please add the missing Copyright section.

the bit as a boolean.

Take, for example, a height with binary encoding `0010 0111`. To encode a coinbase output at this height, one begins
with a left shift: `0100 1110`, and places the coinbase flag in the least significant bit: `0100 1111`.

#### Reconstructable Script Format

Scripts are serialized in this format by concatenating the `Prefix` and `Format` fields specified below.

| Prefix | Script   | Format                                | Expansion                                                    |
| :----- | :------- | :------------------------------------ | :----------------------------------------------------------- |
| `0x00` | Unknown  | `CompactSize(Len([]bytes)) + []bytes` | `[]bytes`                                                    |
| `0x01` | `P2PKH`  | `<20 bytes>`                          | `OP_DUP OP_HASH160 20 <20 bytes> OP_EQUALVERIFY OP_CHECKSIG` |
| `0x02` | `P2PK`   | `<32-byte public key (0x02 parity)>`  | `33 0x02 <32 byte public key> OP_CHECKSIG`                   |
| `0x03` | `P2PK`   | `<32-byte public key (0x03 parity)>`  | `33 0x03 <32 byte public key> OP_CHECKSIG`                   |

> | `0x04` | `P2PK`   | `<64 byte public key>`                | `65 0x04 <64 byte public key> OP_CHECKSIG`                   |

Reviewer 2: Given that they are pretty uncommon these days, and it only saves two bytes, have you considered grouping the P2PK outputs with the other bare scripts and just encoding these per the `0x00` prefix?

| Input index              | 32-bit unsigned integer       | Little endian                       | The index in the block inputs, coinbase excluded. |
| Height and coinbase flag | Height code                   | Defined above                       | —                                                 |
| Script                   | Reconstructable script format | Defined above                       | —                                                 |
| Amount                   | 64-bit unsigned integer       | `CompressAmount` then `CompactSize` | Satoshi-denominated value.                        |

### Messages

#### MSG_GET_SPENT_COINS

`MSG_GET_SPENT_COINS` defines a request for the inputs of a block.

Define `cmdString` as `getbspent`. Define BIP-324 message type as ???.

| Field       | Type                    | Description                                                          |
| :---------- | :---------------------- | :------------------------------------------------------------------- |
| `blockhash` | `<32 bytes>`            | Hash of the block for which inputs are requested.                    |

> | `cutoff`    | 32-bit unsigned integer | If greater than zero, include only coins created before this height. |

Reviewer 2: I could see coins created after a specific height (due to having already processed all blocks up to that height), but why only coins created before a height? Perhaps add an explanation what the `cutoff` is useful for?

Author: https://github.com/bitcoin/bips/pull/2152#discussion_r3486014031

supporting `bspent` MUST adhere to the format of `Coin` specified in the `version` of the request.

## Signaling

Support for serving historical block spent coins is advertised by a feature message, introduced by
[BIP-434](https://github.com/bitcoin/bips/blob/master/bip-0434.md).

| featureid           | featuredata |
| :------------------ | :---------- |
| `blockspentcoinsv1` | `0x00`      |

A client advertising this service bit SHOULD respond to `getbspent` messages, subject to rate-limiting and bandwidth
limiting.

## Rationale

The lifetime, or interval between creation and spending height, of the coins on the Bitcoin blockchain demonstrate an
empirical phenomena that the majority of coins are spent within 100 blocks. In fact, approximately 41 percent of coins
are spent within 10 blocks at the time of writing[^1]. Clients may leverage this to reduce the bandwidth required to
fetch spent coins by using an in-memory cache. For example, a client may store coins that were created in a 5 block
window, and request only coins that are older than this height via the `cutoff` filter. This results in a significant

> bandwidth reduction at the cost of a cache that can be set dynamically by the client depending on available memory.

Reviewer 2: Ah cool. I was missing this context above when `cutoff` was introduced.

Author: I added a short note when introducing the request message that the `cutoff` field is motivated in the rationale section.

```
  BIP: ?
  Layer: Peer Services

>   Title: Peer sharing of block spent coins

Reviewer 3: nit: the naming is inconsistent in these files. Some places use the blockundo terminology and some (like featureid blockspent)

Author: Grepped and removed all "block undo" omitting the nomenclature explanation

Define `cmdString` as `bspent`. Define BIP-324 message type as ???.

| Field       | Type                             | Description                                       |
| :---------- | :------------------------------- | :------------------------------------------------ |
| `blockhash` | `<32 bytes>`                     | Block hash these coins are spent from.            |
| `len`       | `CompactSize(Len(vector<Coin>))` | Length of the coins vector.                       |
| `coins`     | `vector<Coin>`                   | Coins spent, after filtering on request `cutoff`. |

A client supporting the `bspent` MUST include coins created _before_ the `cutoff` field in `getbspent` requests. A
client receiving a `bspent` message with un-requested or missing coins MUST disconnect from the serving peer. A client
supporting `bspent` MUST adhere to the format of `Coin` specified in the `version` of the request.

## Signaling

Support for serving historical block undo data is advertised by a feature message, introduced by
[BIP-434](https://github.com/bitcoin/bips/blob/master/bip-0434.md).

| featureid           | featuredata |
| :------------------ | :---------- |

> | `blockspentcoinsv1` | `0x00`      |

Reviewer 3: nit: this version in feature name may confuse readers with encoding version described for `Coin`.

Author: Not sure I follow. I think `Coin` is only described as a serialization format but does not specify a version AFICT

Author: Oh got it. The version supported by the server is now implied by `featuredata`, so the `MSG_GET_SPENT_COINS` version can be omitted entirely.

A client advertising this service bit SHOULD respond to `getbspent` messages, subject to rate-limiting and bandwidth
limiting.

## Rationale

The lifetime, or interval between creation and spending height, of the coins on the Bitcoin blockchain demonstrate an
empirical phenomena that the majority of coins are spent within 100 blocks. In fact, approximately 41 percent of coins
are spent within 10 blocks at the time of writing[^1]. Clients may leverage this to reduce the bandwidth required to
fetch undo data by using an in-memory cache. For example, a client may store coins that were created in a 5 block
window, and request only coins that are older than this height via the `cutoff` filter. This results in a significant

> bandwidth reduction at the cost of a cache that can be set dynamically by the client depending on available memory.

Reviewer 3: This cutoff cache optimization seems to nudge implementors back to sequentially processing blocks with the added burden of requesting extra data over the wire.

Also with the current messages I still need to get the data for the block (even if there's only one unspent cache miss?), right?
If that's true, wouldn't a parameter for inputs of interest (delta encoded index) help here?

At `150ms RTT * 955233` blocks that's `~39.8 hrs` of round-trip latency for uncached requests alone, before counting download time (as Roasbeef noted in an earlier review). It seems to me that the cache mitigates this but at the cost of reintroducing the very sequentiality it aims to eliminate.

Is this understanding correct?

Reviewer 4: >seems to nudge implementors back to sequentially processing blocks

Caching requires sequential processing, but you can have multiple sequential threads in parallel.

>added burden of requesting extra data over the wire

You're going to have to request the undo data regardless for non-assumevalid SwiftSync - it is not related to caching.

>I still need to get the data for the block (even if there's only one unspent cache miss?), right?

>It seems to me that the cache mitigates [round-trip latency]

Caching does not prevent the need for requesting undo data. You can safely assume pretty much every block has cache misses. No cache missses is equivalent to having the full UTXO set (and impossible with multiple sequential threads), which defeats the point.

>round-trip latency for uncached requests

I have no strong opinion on batching, but round-trip latency won't add up sequentially if requests are sent out in parallel.

Concrete example: Let's say you're starting another sequential thread from block height 1001 and you intend to cache the last 5 blocks worth of outputs. For the first block you'd request the full undo data. For block 1002 until 1005 you'd request everything created until block height 1000. From height 1006 onwards your 5-block window starts to shift so you'd request everything created until block height 1001, and so on.

All this data can be requested in parallel. As long as your caching strategy is not based on what you witnessed during the previous block, at no point do you have to wait for one block to finish processing before requesting the data for upcoming blocks.

and amount metadata for each spend transaction output of a block are stored in a data structure known colloquially as
"undo data". This terminology stems from its use to "undo" the effect of a block by repopulating the UTXO set with the
coins that were spent by the reorganized block. To remain general in language, this data will be referred as "spent
coins."

Bitcoin Core full archival nodes store spent coins for all blocks. This is useful in the context of SwiftSync, as no
additional index must be created or maintained to serve this data to peers. There are, however, some discrepancies
between how this data is serialized on disk in Bitcoin Core and how this proposal seeks to serialize this data over the
peer-to-peer protocol, which are detailed in the rationale section.

This section defines how to request and serve block spent coins over the peer-to-peer protocol, as well as signaling
support of this feature to peers.

### Definitions

- `[]byte`: arbitrary sequence of bytes with no fixed length
- `<N bytes>`: byte vector of size N, where N is specified inline. N is fixed length and known at compile time (e.g.

>   \<32 bytes>)

Reviewer 2: errant backslash, did you mean:

```suggestion
- `<N bytes>`: byte vector of size N, where N is specified inline. N is fixed length and known at compile time (e.g.
  `<32 bytes>`)
```

- `vector<Foo>`: vector of arbitrary length of elements of type Foo
- `CompactSize`: encoding of unsigned integers defined in peer-to-peer messages, as defined in the Function Appendix
  section
- `CompressAmount`: compression function for integer amounts, as defined in the Function Appendix section

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and
"OPTIONAL" in this document are to be interpreted as described in RFC 2119.

### Data structures

#### Height and Coinbase Flag Code

When validating a block, a client must confirm coinbase outputs are mature, which is given by the height of the coin.

> The height and coinbase flag are encoded as a 32 bit integer[^1]. To encode the height and flag, binary left shift the

Reviewer 2: ```suggestion
The height and coinbase flag are encoded as a 32-bit integer[^1]. To encode the height and flag, binary left shift the
```

| `0x00` | Unknown  | `CompactSize(Len([]bytes)) + []bytes` | `[]bytes`                                                    |
| `0x01` | `P2PKH`  | `<20 bytes>`                          | `OP_DUP OP_HASH160 20 <20 bytes> OP_EQUALVERIFY OP_CHECKSIG` |
| `0x02` | `P2PK`   | `<32-byte public key (0x02 parity)>`  | `33 0x02 <32 byte public key> OP_CHECKSIG`                   |
| `0x03` | `P2PK`   | `<32-byte public key (0x03 parity)>`  | `33 0x03 <32 byte public key> OP_CHECKSIG`                   |
| `0x04` | `P2PK`   | `<64 byte public key>`                | `65 0x04 <64 byte public key> OP_CHECKSIG`                   |
| `0x05` | `P2SH`   | `<20 bytes>`                          | `OP_HASH160 20 <20 bytes> OP_EQUAL`                          |
| `0x06` | `P2WSH`  | `<32 bytes>`                          | `OP_0 32 <32 bytes>`                                         |
| `0x07` | `P2WPKH` | `<20 bytes>`                          | `OP_0 20 <20 bytes>`                                         |
| `0x08` | `P2TR`   | `<32-byte X-only public key>`         | `OP_1 32 <32 bytes>`                                         |

#### Amount Format

> The 64 bit unsigned integers representing amounts are compressed by first using the `CompressAmount` function defined

Reviewer 2: ```suggestion
The 64-bit unsigned integers representing amounts are compressed by first using the `CompressAmount` function defined
```

below, and serializing the result with `CompactSize`.

#### Coin

| Field                    | Type                          | Serialization                       | Description                                       |
| :----------------------- | :---------------------------- | :---------------------------------- | :------------------------------------------------ |
| Input index              | 32-bit unsigned integer       | Little endian                       | The index in the block inputs, coinbase excluded. |
| Height and coinbase flag | Height + Coinbase Flag Code   | Defined above                       | —                                                 |
| Script                   | Reconstructable script format | Defined above                       | —                                                 |
| Amount                   | 64-bit unsigned integer       | `CompressAmount` then `CompactSize` | Satoshi-denominated value.                        |

### Messages

#### MSG_GET_SPENT_COINS

`MSG_GET_SPENT_COINS` defines a request for the inputs of a block.

> Define `cmdString` as `getbspent`. Define BIP-324 message type as ???.

Reviewer 5: We lack a message type here.

| Field       | Type                    | Description                                                          |
| :---------- | :---------------------- | :------------------------------------------------------------------- |
| `blockhash` | `<32 bytes>`            | Hash of the block for which inputs are requested.                    |
| `cutoff`    | 32-bit unsigned integer | If greater than zero, include only coins created before this height. |

Rationale of the `cutoff` field is detailed in the rationale section below.

#### MSG_SPENT_COINS

`MSG_SPENT_COINS` defines the data structure for inputs of a block.

> Define `cmdString` as `bspent`. Define BIP-324 message type as ???.

Reviewer 5: ...and here.

| Field       | Type                             | Description                                       |
| :---------- | :------------------------------- | :------------------------------------------------ |
| `blockhash` | `<32 bytes>`                     | Block hash these coins are spent from.            |
| `len`       | `CompactSize(Len(vector<Coin>))` | Length of the coins vector.                       |
| `coins`     | `vector<Coin>`                   | Coins spent, after filtering on request `cutoff`. |

> A client supporting the `bspent` MUST include coins created _before_ the `cutoff` field in `getbspent` requests. A

Reviewer 2: I think there is a word missing here. Is it maybe message?
```suggestion
A client supporting the `bspent` message MUST include coins created _before_ the `cutoff` field in `getbspent` requests. A
```

A client advertising this feature SHOULD respond to `getbspent` messages, subject to rate-limiting and bandwidth
limiting.

## Rationale

The lifetime, or interval between creation and spending height, of the coins on the Bitcoin blockchain demonstrate an
empirical phenomena that the majority of coins are spent within 100 blocks. In fact, approximately 41 percent of coins
are spent within 10 blocks at the time of writing[^2]. Clients may leverage this to reduce the bandwidth required to
fetch spent coins by using an in-memory cache. For example, a client may store coins that were created in a 5 block
window, and request only coins that are older than this height via the `cutoff` filter. This results in a significant
bandwidth reduction at the cost of a cache that can be set dynamically by the client depending on available memory.

Beyond the use of a dynamic coin height filter, there are additional reasons to not simply read the spent coins from
disk and send it over the wire. Legacy fields (`nVersion`) are set to `0x00` when writing and reading this data to
maintain compatibility of disk format with old clients. Furthermore, using the amount compression specified above, an

> 11gb reduction in bandwidth is achieved. `CompactSize`, which is commonly used in P2P messages to describe collection

Reviewer 2: Please use either GB for gigabyte or GiB for gibibyte and add a space between number and unit symbol.

```suggestion
11 [GB|GiB] reduction in bandwidth is achieved. `CompactSize`, which is commonly used in P2P messages to describe collection
```

> lengths, was selected over `VARINT`, which is used internally within Bitcoin Core to represent variable length integers

Reviewer 2: ```suggestion
lengths, was selected over `VARINT`, which is used internally within Bitcoin Core to represent variable length integers.
```

> The application of `VARINT` as opposed to `CompactSize` offers a further reduction of 4gb, however the `VARINT`

Reviewer 2: ```suggestion
The application of `VARINT` as opposed to `CompactSize` offers a further reduction of 4 [GB|GiB], however the `VARINT`
```

primitive is currently a Bitcoin Core implementation detail and has not been included in a P2P message. Reusing existing
network primitives results in the majority of savings, so this specification opts to lower implementation burden for

> clients. With respect to reconstructable script, utilizing this format results in a savings of around 12gb. The scheme

Reviewer 2: ```suggestion
clients. With respect to reconstructable script, utilizing this format results in a savings of around 12 [GB|GiB]. The scheme
```

is loss-less, and may be upgradable by appending script variants. For reference, the naive encoding of block spent coins

> is 118gb as of block 930,000[^2][^3][^4].

Reviewer 2: Just about a 7% difference. ;)

```suggestion
is 118 [GB|GiB] as of block 930,000[^2][^3][^4].
```

[^1]: When representing objects in memory, programming languages will align the bytes of the fields of an object. A
boolean is commonly padded to 4 or 8 bytes, but only requires a bit. Further, if the coinbase flag was a separate field
as represented in the message, it would require at least one byte. Losing one bit of precision in the block height still
allows for valid encodings of heights up to 2,147,483,647.
[^2]: Relevant statistics may be generated via binaries in
the [`swiftsync-research`](https://github.com/rustaceanrob/swiftsync-research) repository
[^3]: Reconstructable
scripts are borrowed from [UTREEXO](https://github.com/bitcoin/bips/pull/1923) which is subsequently borrowed from Cory
Field's UHS proposal
[^4]: Astute readers may notice uncompressed public keys may be compressed before they are sent
and decompressed by the receiving client. Although this would slightly reduce bandwidth, it would increase the
complexity of client code, as a `secp256k1` context would be required to decode the message, which is not currently a
requirement. As of height 936,212 the number of uncompressed public keys spent in blocks is 853,515. This represents a

> very modest savings in bandwidth, around 30MB. As such, this technique is omitted for implementation simplicity.

Reviewer 2: ```suggestion
very modest savings in bandwidth, around 30 MB. As such, this technique is omitted for implementation simplicity.
```
